### PR TITLE
Milestonebuilds

### PR DESCRIPTION
**Pull request by cniweb**
Add Milestone Builds #192

---
This change is [Reviewable](https://reviewable.io/reviews/openhab/openhab-docker/193)

### CODE REVIEW

File: 2.4.0.M3/i386/debian/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+# Copy initial files to host volume
+case ${OPENHAB_VERSION} in
+  1.8.3)
+      if [ -z "$(ls -A "${APPDIR}/configurations")" ]; then
+        # Copy userdata dir for version 1.8.3
+        echo "No configuration found... initializing."
+        cp -av "${APPDIR}/configurations.dist/." "${APPDIR}/configurations/"
+      fi
+    ;;
+  2.0.0|2.1.0|2.2.0|2.3.0|2.4.0-snapshot)
```
Comment (wborn) [inline 2.4.0.M3/i386/debian/entrypoint.sh]: The milestones are missing here.

File: update.sh [hunk 2]
```diff
@@ -303,7 +312,7 @@ print_expose_ports() {
 		expose_comment="Expose HTTP, HTTPS and Console ports"
 		expose_ports="8080 8443 8101"
 		;;
-	2.2.0|2.3.0|2.4.0-snapshot)
+	2.2.0|2.3.0|2.4.0-snapshot|2.4.0.M1|2.4.0.M2|2.4.0.M3)
```
Comment (wborn) [inline update.sh]: According to the [community post](https://community.openhab.org/t/openhab-2-4-milestone-builds/50359) the first usable milestone was M3. So I think noone should ever use the M1 and M2 milestone builds.

File: 2.4.0.M2/arm64/alpine/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# openhab image
+#
+# ------------------------------------------------------------------------------
+#               NOTE: THIS DOCKERFILE IS GENERATED VIA "update.sh"
+#
+#                       PLEASE DO NOT EDIT IT DIRECTLY.
+# ------------------------------------------------------------------------------
+#
+FROM multiarch/alpine:arm64-v3.7
```
Comment (wborn) [inline 2.4.0.M2/arm64/alpine/Dockerfile]: Looks like you are missing some commits in your branch because a PR was merged to upgrade to alpine 3.8.

File: 2.4.0.M2/arm64/alpine/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+# Set download urls
+ENV 	    JAVA_URL="https://www.azul.com/downloads/zulu/zdk-8-ga-linux_aarch32hf.tar.gz" 	    OPENHAB_URL="https://openhab.jfrog.io/openhab/libs-milestone-local/org/openhab/distro/openhab/2.4.0.M2/openhab-2.4.0.M2.tar.gz" 	    OPENHAB_VERSION="2.4.0.M2"
```
Comment (wborn) [inline 2.4.0.M2/arm64/alpine/Dockerfile]: There was also a commit to nicely layout these ENV lines which seems to be missing from your branch.

File: 2.4.0.M2/arm64/alpine/Dockerfile (hunk 3 above)
Comment (cniweb) [reply 2.4.0.M2/arm64/alpine/Dockerfile]: OK, I update to 3.8

File: update.sh (hunk 2 above)
Comment (cniweb) [reply update.sh]: You're right. I have removed M1 and M2.

File: 2.4.0.M3/i386/debian/entrypoint.sh (hunk 1 above)
Comment (cniweb) [reply 2.4.0.M3/i386/debian/entrypoint.sh]: OK, I added.

File: 2.0.0/amd64/debian/Dockerfile [hunk 5]
```diff
@@ -112,3 +112,120 @@ ENTRYPOINT ["/entrypoint.sh"]
...
+    && dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')" \
+    && wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch" \
+    && wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc" \
+    && export GNUPGHOME \
+    && GNUPGHOME="$(mktemp -d)" \
+    && GPG_KEY="B42F6819007F00F88E364FD4036A9C25BF357DD4" \
+    && gpg --keyserver ha.pool.sks-keyservers.net --recv-keys $GPG_KEY \
+       || gpg --keyserver pgp.mit.edu --recv-keys $GPG_KEY \
+       || gpg --keyserver keyserver.pgp.com --recv-keys $GPG_KEY \
+    && gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu \
```
Comment (wborn) [inline 2.0.0/amd64/debian/Dockerfile]: All this gosu code was removed with https://github.com/openhab/openhab-docker/pull/180

Comment (wborn) [reply 2.0.0/amd64/debian/Dockerfile]: Perhaps it's better to create a branch from the master and apply your changes there?

File: 2.4.0.M2/arm64/alpine/Dockerfile (hunk 4 above)
Comment (cniweb) [reply 2.4.0.M2/arm64/alpine/Dockerfile]: Fixed

File: 2.0.0/amd64/debian/Dockerfile (hunk 5 above)
Comment (cniweb) [reply 2.0.0/amd64/debian/Dockerfile]: OK, up update from upstream/master

File: update.sh [hunk 6]
```diff
@@ -82,7 +85,7 @@ print_baseimage() {
 	    OPENHAB_URL="$openhab_url" \\
 	    OPENHAB_VERSION="$version"
 
-	EOI
+EOI
```
Comment (wborn) [inline update.sh]: Can you revert the changes on lines 79 and 88? The generated Dockerfiles now contain `\\` and $base_image instead of the variable value.

File: 2.0.0/amd64/debian/Dockerfile (hunk 5 above)
Comment (wborn) [reply 2.0.0/amd64/debian/Dockerfile]: Thanks. This looks a lot better! :smile:

File: entrypoint_debian.sh [hunk 7]
```diff
@@ -54,7 +54,7 @@ case ${OPENHAB_VERSION} in
         cp -av "${APPDIR}/configurations.dist/." "${APPDIR}/configurations/"
       fi
     ;;
-  2.0.0|2.1.0|2.2.0|2.3.0|2.4.0-snapshot)
+  2.2.0|2.3.0|2.4.0-snapshot|2.4.0.M3)
```
Comment (wborn) [inline entrypoint_debian.sh]: `2.0.0|2.1.0` seems to be missing now

File: entrypoint_alpine.sh [hunk 8]
```diff
@@ -42,7 +42,7 @@ case ${OPENHAB_VERSION} in
         cp -av "${APPDIR}/configurations.dist/." "${APPDIR}/configurations/"
       fi
     ;;
-  2.0.0|2.1.0|2.2.0|2.3.0|2.4.0-snapshot)
+  2.2.0|2.3.0|2.4.0-snapshot|2.4.0.M3)
```
Comment (wborn) [inline entrypoint_alpine.sh]: `2.0.0|2.1.0` seems to be missing now

File: update.sh [hunk 9]
```diff
@@ -17,13 +17,16 @@ print_header() {
 	#                       PLEASE DO NOT EDIT IT DIRECTLY.
 	# ------------------------------------------------------------------------------
 	#
-	EOI
+EOI
```
Comment (wborn) [inline update.sh]: Why change the indentation of lines 11 and 20? The other methods use the same kind of indentation.

File: update.sh [hunk 10]
```diff
@@ -299,7 +302,7 @@ print_expose_ports() {
 	cat >> $1 <<-EOI
 	# $expose_comment
 	EXPOSE $expose_ports
-
+
```
Comment (wborn) [inline update.sh]: This tab is not used so I think it can be removed.

File: update.sh (hunk 6 above)
Comment (cniweb) [reply update.sh]: Fixed

File: entrypoint_debian.sh (hunk 7 above)
Comment (cniweb) [reply entrypoint_debian.sh]: Fixed

File: entrypoint_alpine.sh (hunk 8 above)
Comment (cniweb) [reply entrypoint_alpine.sh]: Fixed

File: update.sh (hunk 10 above)
Comment (cniweb) [reply update.sh]: But the Dockerfile looks nicely formatted.

File: update.sh (hunk 9 above)
Comment (cniweb) [reply update.sh]: Fixed

File: update.sh [hunk 11]
```diff
@@ -2,13 +2,13 @@
...
 # Dockerfiles to be generated
-versions="2.4.0-snapshot 2.3.0 2.2.0 2.1.0 2.0.0 1.8.3"
+versions="2.4.0.M3 2.4.0-snapshot 2.3.0 2.2.0 2.1.0 2.0.0 1.8.3"
 arches="i386 amd64 armhf arm64"
 bases="debian alpine"
 
 # Generate header
 print_header() {
-	cat > $1 <<-EOI
+    cat > $1 <<-EOI
```
Comment (wborn) [inline update.sh]: Code wise I think it's now all OK!
I'd just replace these 4 spaces with a tab so it is similar to the other methods. ;-)

File: update.sh [hunk 12]
```diff
@@ -299,8 +302,8 @@ print_expose_ports() {
 	cat >> $1 <<-EOI
 	# $expose_comment
 	EXPOSE $expose_ports
-
-	EOI
+
```
Comment (wborn) [inline update.sh]: And remove this tab because it's not used for anything.

**Comment by wborn** [comment]
I think the milestone build also needs to be added to the matrix in .travis.yml

File: update.sh (hunk 11 above)
Comment (cniweb) [reply update.sh]: OK, fixed

File: update.sh (hunk 12 above)
Comment (cniweb) [reply update.sh]: Fixed

**Comment by wborn** [comment]
Can you also update the documentation?

**Comment by cniweb** [comment]
Travis matrix and documentation changed.

**Comment by cniweb** [comment]
Closes #192

**Comment by CWempe** [comment]
Maybe I don't know enough about all this, but I am having to questions:

* should the milestone tag(s) be already visible at https://hub.docker.com/r/openhab/openhab/tags/ ?
* I would have expected a tag like this: `2.4.0-milestone` that gets automatically updated with each new milestone.
  Just like `-snapshot` gets updated every night.
  Not the current tag `2.4.0.M3` which will not be updated to `M4`.

**Comment by wborn** [comment]
> I would have expected a tag like this: 2.4.0-milestone that gets automatically updated with each new milestone.
> Just like -snapshot gets updated every night.
> Not the current tag 2.4.0.M3 which will not be updated to M4.

Milestone users prefer stability so the unique tags should be there (M3, M4 etc.).

It would be nice to have some more rolling tags for testing though but this also applies to the other image tags we have.  E.g. there's currently no version (2.3, 2.4) independent `stable` or `snapshot` tag as well.

For a good read on Docker tagging best practices see also: https://blogs.msdn.microsoft.com/stevelasker/2018/03/01/docker-tagging-best-practices-for-tagging-and-versioning-docker-images/

If these rolling tags would be introduced I think it would be best to first decrease the current number of tags by using  multi architecture manifests that include the OS and architecture, for more on that see:

* https://developer.ibm.com/linuxonpower/2017/07/27/create-multi-architecture-docker-image/
* https://docs.docker.com/registry/spec/manifest-v2-2/
